### PR TITLE
Return errors from `Dlopen`, `Dlsym`, and `Dlclose`

### LINKED ISSUE #100: Return errors from `Dlerror()`
It would be better if `Dlerror()` returned errors, rather than strings as it does now. That would allow users to check errors using a nil check, rather than an empty string check, as well as allow them to return an error from `purego` as an `error` type from their own function. It would also provide the ability to wrap `purego` errors, such as with `%w` in `fmt.Errorf()`, as well as with the new functions added in Go 1.20, such as `errors.Join()`.

`Dlerror()` could simply use `errors.New()` to create a new error with the given string, but I feel it would be even better to define a custom error type that just returns a string. That way, the type information tells the caller that the error came from `purego`.

This would look something like:

```go
type Error struct {
	s string
}

func (e Error) Error() string {
	return e.s
}
```

If this seems like a good idea, I could submit a PR to implement it.

### PR DESCRIPTION
**Pull request by Arsen6331**
This PR returns errors from the `Dlopen`, `Dlsym`, and `Dlclose` functions in order to provide a more idiomatic API.

Fixes #100

### CODE REVIEW

File: README.md
Comment (hajimehoshi) [inline README.md:49]: `if err := purego.RegisterLibFunc(&puts, libc, "puts"); err != nil`

File: objc/objc_runtime_darwin.go
Comment (hajimehoshi) [inline objc/objc_runtime_darwin.go:56]: ditto

File: error.go
Comment (hajimehoshi) [inline error.go:4]: @TotallyGamerJet Don't we need to distinguish libdl errors or others? If so, we might want a special type for libdl error

File: objc/objc_runtime_darwin.go
Comment (Arsen6331) [reply objc/objc_runtime_darwin.go]: Done

File: error.go
Comment (Arsen6331) [reply error.go:4]: I added this error type specifically for libdl errors, so I it should already be distinguished from other errors. I can make the name more specific, like `LibdlError`, if you think that would be better.

File: dlfcn.go
Comment (hajimehoshi) [inline dlfcn.go:39]:
```go
if errStr := fnDlerror(); errStr != "" {
```

File: dlfcn.go
Comment (hajimehoshi) [inline dlfcn.go:52]: ditto

File: dlfcn.go
Comment (hajimehoshi) [inline dlfcn.go:64]: I feel like we can remove this function instead of deprecating.

@TotallyGamerJet, thoughts?

File: error.go
Comment (hajimehoshi) [inline error.go:1]: Add a license comment at the top of this file

```go
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: 2023 The Ebitengine Authors
```

File: dlfcn.go
Comment (Arsen6331) [reply dlfcn.go]: Done

File: error.go
Comment (hajimehoshi) [reply error.go:7]: `DlError` sounds better (if we decide to remove the `Dlerror` function)

File: func.go
Comment (hajimehoshi) [inline func.go:24]: Hmm, I'm not confident whether we should return an error here...

File: error.go
Comment (TotallyGamerJet) [reply error.go:7]: I saw the comment above

Comment (TotallyGamerJet) [reply error.go:7]: I think Dlerror sounds good

File: func.go
Comment (hajimehoshi) [reply func.go:21]: > Should this still panic instead of return an error? 

I agree.

File: dlfcn.go
Comment (Arsen6331) [reply dlfcn.go]: Done

File: objc/objc_runtime_darwin.go
Comment (TotallyGamerJet) [inline objc/objc_runtime_darwin.go:52]: @hajimehoshi I think this error should be wrapped. Do you agree?

File: error.go
Comment (TotallyGamerJet) [inline error.go:6]: This error isn't a generic purego error. It only occurs when using dlopen, dlsym, or dlclose. I think the doc should express that

**Review by TotallyGamerJet — CHANGES_REQUESTED** [review]

File: objc/objc_runtime_darwin.go
Comment (hajimehoshi) [reply objc/objc_runtime_darwin.go:52]: I don't understand. Isn't err already a DlError?

**Comment by hajimehoshi** [comment]
Please check the test results.

File: objc/objc_runtime_darwin.go
Comment (TotallyGamerJet) [reply objc/objc_runtime_darwin.go:52]: Yes but if we wanted the context that this error happens in package objc. Otherwise, just panic with the error

**Comment by Arsen6331** [comment]
> Please check the test results.

The `examples` directory it mentions doesn't exist in my PR, it was only added to main yesterday, along with the `libc` example.

**Comment by hajimehoshi** [comment]
> The examples directory it mentions doesn't exist in my PR, it was only added to main yesterday, along with the libc example.

OK so please rebase or merge your PR and fix the example to pass the test.

File: objc/objc_runtime_darwin.go
Comment (hajimehoshi) [reply objc/objc_runtime_darwin.go:52]: I don't understand your suggestion yet. Are you suggesting to create a new error type for Objc? Even if so, this line would not change, right?

**Comment by Arsen6331** [comment]
> OK so please rebase or merge your PR and fix the example to pass the test.

Done

**Comment by Arsen6331** [comment]
Fixed the failing test

**Review by TotallyGamerJet — COMMENTED** [review]
LGTM!

I'll let @hajimehoshi take one last look before merge.

**Review by hajimehoshi — APPROVED** [review]
LGTM!

File: objc/objc_runtime_darwin.go
Comment (hajimehoshi) [reply objc/objc_runtime_darwin.go]: Was this resolved?

image

**Review by TotallyGamerJet — APPROVED** [review]
LGTM!

**Comment by hajimehoshi** [comment]
As this is a breaking change,

* I tagged the latest commit `0befdc3778764be7c9997dbe0326b2761eeafcd1` as `v0.3.0-alpha`
* I tagged the previous commit `0c858ac9f9ff921a4d24b833c0354cda7fdafa1b` as `v0.2`
* I created a branch from `v0.2`.

**Pull request merged**
